Source/WebKit/UIProcess/Inspector/WebPageDebuggable.cpp

```diff
@@ -39,59 +40,93 @@ using namespace Inspector;
+    : m_page(page)
 {
-    if (!m_page->mainFrame())
-        return String();
-
-    return m_page->mainFrame()->title();
 }
 
-String WebPageDebuggable::url() const
+void WebPageDebuggable::detachFromPage()
 {
-    if (!m_page->mainFrame())
-        return aboutBlankURL().string();
+    m_page = nullptr;
+}
 
-    String url = m_page->mainFrame()->url().string();
-    if (url.isEmpty())
-        return aboutBlankURL().string();
+WebPageDebuggable::~WebPageDebuggable() = default;
 
-    return url;
+String WebPageDebuggable::name() const
+{
+    String name;
+    callOnMainRunLoopAndWait([this, protectedThis = Ref { *this }, &name] {
+        RefPtr page = m_page.get();
+        if (!page || !page->mainFrame())
+            return;
+        name = page->mainFrame()->title().isolatedCopy();
+    });
+    return name;
 }
 
-Ref<WebPageProxy> WebPageDebuggable::protectedPage() const
+
+String WebPageDebuggable::url() const
 {
-    return m_page.get();
+    String url;
+    callOnMainRunLoopAndWait([this, protectedThis = Ref { *this }, &url] {
+        RefPtr page = m_page.get();
+        if (!page || !page->mainFrame()) {
+            url = aboutBlankURL().string().isolatedCopy();
+            return;
+        }
+
+        url = page->mainFrame()->url().string().isolatedCopy();
+        if (url.isEmpty())
+            url = aboutBlankURL().string().isolatedCopy();
+    });
+    return url;
 }
 
 bool WebPageDebuggable::hasLocalDebugger() const
 {
-    return protectedPage()->inspectorController().hasLocalFrontend();
+    bool hasLocalDebugger;
+    callOnMainRunLoopAndWait([this, protectedThis = Ref { *this }, &hasLocalDebugger] {
+        RefPtr page = m_page.get();
+        hasLocalDebugger = page && page->inspectorController().hasLocalFrontend();
+    });
+    return hasLocalDebugger;
 }
 
 void WebPageDebuggable::connect(FrontendChannel& channel, bool isAutomaticConnection, bool immediatelyPause)
 {
-    protectedPage()->inspectorController().connectFrontend(channel, isAutomaticConnection, immediatelyPause);
+    callOnMainRunLoopAndWait([this, protectedThis = Ref { *this }, &channel, isAutomaticConnection, immediatelyPause] {
+        if (RefPtr page = m_page.get())
+            page->inspectorController().connectFrontend(channel, isAutomaticConnection, immediatelyPause);
+    });
 }
 
 void WebPageDebuggable::disconnect(FrontendChannel& channel)
 {
-    protectedPage()->inspectorController().disconnectFrontend(channel);
+    callOnMainRunLoopAndWait([this, protectedThis = Ref { *this }, &channel] {
+        if (RefPtr page = m_page.get())
+            page->inspectorController().disconnectFrontend(channel);
+    });
 }
 
 void WebPageDebuggable::dispatchMessageFromRemote(String&& message)
 {
-    protectedPage()->inspectorController().dispatchMessageFromFrontend(WTFMove(message));
+    callOnMainRunLoopAndWait([this, protectedThis = Ref { *this }, message = WTFMove(message).isolatedCopy()] {
```

Comment:
`callOnMainRunLoopAndWait([this, protectedThis = Ref { *this }, message = WTFMove(message).isolatedCopy()]() mutable {`